docs/features/patterns.md

```diff
@@ -0,0 +1,472 @@
+Pattern Matching for C#
+=======================
+
+Pattern matching extensions for C# enable many of the benefits of algebraic data types and pattern matching from functional languages, but in a way that smoothly integrates with the feel of the underlying language. The basic features are: [record types](records.md), which are types whose semantic meaning is described by the shape of the data; and pattern matching, which is a new expression form that enables extremely concise multilevel decomposition of these data types. Elements of this approach are inspired by related features in the programming languages [F#](http://www.msr-waypoint.net/pubs/79947/p29-syme.pdf "Extensible Pattern Matching Via a Lightweight Language") and [Scala](http://lampwww.epfl.ch/~emir/written/MatchingObjectsWithPatterns-TR.pdf "Matching Objects With Patterns").
+
+## Is Expression
+
+The `is` operator is extended to test an expression against a *pattern*.
+
+>*relational-expression*:
+&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;*relational-expression* `is` *complex-pattern*
+&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;*relational-expression* `is` *type*
+
+It is a compile-time error if *e* does not designate a value or does not have a type.
+
+Every *identifier* of the pattern introduces a new local variable that is *definitely assigned* after the `is` operator is `true` (i.e. *definitely assigned when true*).
+
+## Patterns
+
+Patterns are used in the `is` operator and in a *switch-statement* to express the shape of data against which incoming data is to be compared. Patterns may be recursive so that subparts of the data may be matched against subpatterns.
+
+>*complex-pattern*:
+&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;*type* *identifier*
```

Comment:
Could be the *identifier* optional so that `case int` would be possible as in `if(obj is int)`?